compiler/rustc_errors/src/emitter.rs

```diff
@@ -2263,13 +2268,52 @@ impl HumanEmitter {
                     }
                     if let DisplaySuggestion::Diff = show_code_change {
                         // Colorize removal with red in diff format.
-                        buffer.set_style_range(
-                            row_num - 2,
-                            (padding as isize + span_start_pos as isize) as usize,
-                            (padding as isize + span_end_pos as isize) as usize,
-                            Style::Removal,
-                            true,
-                        );
+
+                        let newlines = snippet.lines().count();
+                        if newlines > 0 && row_num > newlines {
+                            // Account for removals where the part being removed spans multiple
+                            // lines.
+                            // FIXME: We check the number of rows because in some cases, like in
+                            // `tests/ui/lint/invalid-nan-comparison-suggestion.rs`, the rendered
+                            // suggestion will only show the first line of code being replaced. The
+                            // proper way of doing this would be to change the suggestion rendering
+                            // logic to show the whole prior snippet, but the current output is not
+                            // too bad to begin with, so we side-step that issue here.
+                            for (i, line) in snippet.lines().enumerate() {
+                                let line = normalize_whitespace(line);
+                                let row = row_num - 2 - (newlines - i - 1);
```

Comment:
Question [OFFSETS 1/2]: what's the significance of `row_num - 2`? There's several `+/- n` in this region of code which are not immediately obvious (modulo the `- 1` cases), but this one in particular is somewhat mysterious (at least to me).

Reply:
`row_num` at this point corresponds to

```
   |
LL | CODE
   | ++++  <- this row
```

in the buffer. When we have a diff format output, we end up with

```
   |
LL - OLDER
LL + NEWER
   |         <- this row
```

The `row_num - 2` is to select the buffer line that has the "old version of the diff" at that point. When the removal is a single line, `i` is `0`, `newlines` is `1` so `(newlines - i - 1)` ends up being `0`, so `row` points at `LL - OLDER`. When the removal corresponds to multiple lines, we end up with `newlines > 1` and `i` being `0..newlines - 1`.

```
   |
LL - OLDER
LL - CODE
LL - BEING
LL - REMOVED
LL + NEWER
   |         <- this row
```

How should I communicate this better in a comment?

Reply:
Honestly? I can't think of any shorter explanation that's also unambiguous. I would just copy-pasta what you said as a comment, even at the risk of the comment being outdated. Because until you explained this to me (thank you), I don't think anyone else would've been able to interpret this without extensive local testing...